## Overview

This notebook will show you how to create and query a table or DataFrame that you uploaded to DBFS. [DBFS](https://docs.databricks.com/user-guide/dbfs-databricks-file-system.html) is a Databricks File System that allows you to store data for querying inside of Databricks. This notebook assumes that you have a file already inside of DBFS that you would like to read from.

This notebook is written in **Python** so the default cell type is Python. However, you can use different languages by using the `%LANGUAGE` syntax. Python, Scala, SQL, and R are all supported.

# What are the risk factors for heart disease?

###Import mlflow

In [4]:
dbutils.library.installPyPI("mlflow")
dbutils.library.restartPython()
import mlflow


PyPI package mlflow has been installed already. The previously-installed package is `mlflow`. To resolve this issue detach and re-attach the notebook to create a new environment or rename the package.

### Importing Libraries & Dataset

In [6]:
# File location and type
file_location = "/FileStore/tables/HeartDisease.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)

display(df)

age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target 63 1 3 145 233 1 0 150 0 2.3 0 0 1 1 37 1 2 130 250 0 1 187 0 3.5 0 0 2 1 41 0 1 130 204 0 0 172 0 1.4 2 0 2 1 56 1 1 120 236 0 1 178 0 0.8 2 0 2 1 57 0 0 120 354 0 1 163 1 0.6 2 0 2 1 57 1 0 140 192 0 1 148 0 0.4 1 0 1 1 56 0 1 140 294 0 0 153 0 1.3 1 0 2 1 44 1 1 120 263 0 1 173 0 0.0 2 0 3 1 52 1 2 172 199 1 1 162 0 0.5 2 0 3 1 57 1 2 150 168 0 1 174 0 1.6 2 0 2 1 54 1 0 140 239 0 1 160 0 1.2 2 0 2 1 48 0 2 130 275 0 1 139 0 0.2 2 0 2 1 49 1 1 130 266 0 1 171 0 0.6 2 0 2 1 64 1 3 110 211 0 0 144 1 1.8 1 0 2 1 58 0 3 150 283 1 0 162 0 1.0 2 0 2 1 50 0 2 120 219 0 1 158 0 1.6 1 0 2 1 58 0 2 120 340 0 1 172 0 0.0 2 0 2 1 66 0 3 150 226 0 1 114 0 2.6 0 0 2 1 43 1 0 150 247 0 1 171 0 1.5 2 0 2 1 69 0 3 140 239 0 1 151 0 1.8 2 2 2 1 59 1 0 135 234 0 1 161 0 0.5 1 0 3 1 44 1 2 130 233 0 1 179 1 0.4 2 0 2 1 42 1 0 140 226 0 1 178 0 0.0 2 0 2 1 61 1 2 150 243 1 1 137 1 1.0 1 0 2 1 40 1 3 140 199 0 1 178 1 1.4 2 0 3 1 71 0 1 160 302 0 1 162 0 0.4 2 2 2 1 59 1 2 150 212 1 1 157 0 1.6 2 0 2 1 51 1 2 110 175 0 1 123 0 0.6 2 0 2 1 65 0 2 140 417 1 0 157 0 0.8 2 1 2 1 53 1 2 130 197 1 0 152 0 1.2 0 0 2 1 41 0 1 105 198 0 1 168 0 0.0 2 1 2 1 65 1 0 120 177 0 1 140 0 0.4 2 0 3 1 44 1 1 130 219 0 0 188 0 0.0 2 0 2 1 54 1 2 125 273 0 0 152 0 0.5 0 1 2 1 51 1 3 125 213 0 0 125 1 1.4 2 1 2 1 46 0 2 142 177 0 0 160 1 1.4 0 0 2 1 54 0 2 135 304 1 1 170 0 0.0 2 0 2 1 54 1 2 150 232 0 0 165 0 1.6 2 0 3 1 65 0 2 155 269 0 1 148 0 0.8 2 0 2 1 65 0 2 160 360 0 0 151 0 0.8 2 0 2 1 51 0 2 140 308 0 0 142 0 1.5 2 1 2 1 48 1 1 130 245 0 0 180 0 0.2 1 0 2 1 45 1 0 104 208 0 0 148 1 3.0 1 0 2 1 53 0 0 130 264 0 0 143 0 0.4 1 0 2 1 39 1 2 140 321 0 0 182 0 0.0 2 0 2 1 52 1 1 120 325 0 1 172 0 0.2 2 0 2 1 44 1 2 140 235 0 0 180 0 0.0 2 0 2 1 47 1 2 138 257 0 0 156 0 0.0 2 0 2 1 53 0 2 128 216 0 0 115 0 0.0 2 0 0 1 53 0 0 138 234 0 0 160 0 0.0 2 0 2 1 51 0 2 130 256 0 0 149 0 0.5 2 0 2 1 66 1 0 120 302 0 0 151 0 0.4 1 0 2 1 62 1 2 130 231 0 1 146 0 1.8 1 3 3 1 44 0 2 108 141 0 1 175 0 0.6 1 0 2 1 63 0 2 135 252 0 0 172 0 0.0 2 0 2 1 52 1 1 134 201 0 1 158 0 0.8 2 1 2 1 48 1 0 122 222 0 0 186 0 0.0 2 0 2 1 45 1 0 115 260 0 0 185 0 0.0 2 0 2 1 34 1 3 118 182 0 0 174 0 0.0 2 0 2 1 57 0 0 128 303 0 0 159 0 0.0 2 1 2 1 71 0 2 110 265 1 0 130 0 0.0 2 1 2 1 54 1 1 108 309 0 1 156 0 0.0 2 0 3 1 52 1 3 118 186 0 0 190 0 0.0 1 0 1 1 41 1 1 135 203 0 1 132 0 0.0 1 0 1 1 58 1 2 140 211 1 0 165 0 0.0 2 0 2 1 35 0 0 138 183 0 1 182 0 1.4 2 0 2 1 51 1 2 100 222 0 1 143 1 1.2 1 0 2 1 45 0 1 130 234 0 0 175 0 0.6 1 0 2 1 44 1 1 120 220 0 1 170 0 0.0 2 0 2 1 62 0 0 124 209 0 1 163 0 0.0 2 0 2 1 54 1 2 120 258 0 0 147 0 0.4 1 0 3 1 51 1 2 94 227 0 1 154 1 0.0 2 1 3 1 29 1 1 130 204 0 0 202 0 0.0 2 0 2 1 51 1 0 140 261 0 0 186 1 0.0 2 0 2 1 43 0 2 122 213 0 1 165 0 0.2 1 0 2 1 55 0 1 135 250 0 0 161 0 1.4 1 0 2 1 51 1 2 125 245 1 0 166 0 2.4 1 0 2 1 59 1 1 140 221 0 1 164 1 0.0 2 0 2 1 52 1 1 128 205 1 1 184 0 0.0 2 0 2 1 58 1 2 105 240 0 0 154 1 0.6 1 0 3 1 41 1 2 112 250 0 1 179 0 0.0 2 0 2 1 45 1 1 128 308 0 0 170 0 0.0 2 0 2 1 60 0 2 102 318 0 1 160 0 0.0 2 1 2 1 52 1 3 152 298 1 1 178 0 1.2 1 0 3 1 42 0 0 102 265 0 0 122 0 0.6 1 0 2 1 67 0 2 115 564 0 0 160 0 1.6 1 0 3 1 68 1 2 118 277 0 1 151 0 1.0 2 1 3 1 46 1 1 101 197 1 1 156 0 0.0 2 0 3 1 54 0 2 110 214 0 1 158 0 1.6 1 0 2 1 58 0 0 100 248 0 0 122 0 1.0 1 0 2 1 48 1 2 124 255 1 1 175 0 0.0 2 2 2 1 57 1 0 132 207 0 1 168 1 0.0 2 0 3 1 52 1 2 138 223 0 1 169 0 0.0 2 4 2 1 54 0 1 132 288 1 0 159 1 0.0 2 1 2 1 45 0 1 112 160 0 1 138 0 0.0 1 0 2 1 53 1 0 142 226 0 0 111 1 0.0 2 0 3 1 62 0 0 140 394 0 0 157 0 1.2 1 0 2 1 52 1 0 108 233 1 1 147 0 0.1 2 3 3 1 43 1 2 130 315 0 1 162 0 1.9 2 1 2 1 53 1 2 130 246 1 0 173 0 0.0 2 3 2 1 42 1 3 148 244 0 0 178 0 0.8 2 2 2 1 59 1 3 178 270 0 0 145 0 4.2 0 0 3 1 63 0 1 140 195 0 1 179 0 0.0 2 2 2 1 42 1 2 120 240 1 1 194 0 0.8 0 0 3 1 50 1 2 129 196 0 1 163 0 0.0 2 0 2 1 68 0 2 120 211 0 0 115 0 1.5 1 0 2 1

In [7]:
# Create a view or table

temp_table_name = "HeartDisease_csv"

df.createOrReplaceTempView(temp_table_name)

In [8]:
%sql

/* Query the created temp table in a SQL cell */

select * from `HeartDisease_csv`

age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target 63 1 3 145 233 1 0 150 0 2.3 0 0 1 1 37 1 2 130 250 0 1 187 0 3.5 0 0 2 1 41 0 1 130 204 0 0 172 0 1.4 2 0 2 1 56 1 1 120 236 0 1 178 0 0.8 2 0 2 1 57 0 0 120 354 0 1 163 1 0.6 2 0 2 1 57 1 0 140 192 0 1 148 0 0.4 1 0 1 1 56 0 1 140 294 0 0 153 0 1.3 1 0 2 1 44 1 1 120 263 0 1 173 0 0.0 2 0 3 1 52 1 2 172 199 1 1 162 0 0.5 2 0 3 1 57 1 2 150 168 0 1 174 0 1.6 2 0 2 1 54 1 0 140 239 0 1 160 0 1.2 2 0 2 1 48 0 2 130 275 0 1 139 0 0.2 2 0 2 1 49 1 1 130 266 0 1 171 0 0.6 2 0 2 1 64 1 3 110 211 0 0 144 1 1.8 1 0 2 1 58 0 3 150 283 1 0 162 0 1.0 2 0 2 1 50 0 2 120 219 0 1 158 0 1.6 1 0 2 1 58 0 2 120 340 0 1 172 0 0.0 2 0 2 1 66 0 3 150 226 0 1 114 0 2.6 0 0 2 1 43 1 0 150 247 0 1 171 0 1.5 2 0 2 1 69 0 3 140 239 0 1 151 0 1.8 2 2 2 1 59 1 0 135 234 0 1 161 0 0.5 1 0 3 1 44 1 2 130 233 0 1 179 1 0.4 2 0 2 1 42 1 0 140 226 0 1 178 0 0.0 2 0 2 1 61 1 2 150 243 1 1 137 1 1.0 1 0 2 1 40 1 3 140 199 0 1 178 1 1.4 2 0 3 1 71 0 1 160 302 0 1 162 0 0.4 2 2 2 1 59 1 2 150 212 1 1 157 0 1.6 2 0 2 1 51 1 2 110 175 0 1 123 0 0.6 2 0 2 1 65 0 2 140 417 1 0 157 0 0.8 2 1 2 1 53 1 2 130 197 1 0 152 0 1.2 0 0 2 1 41 0 1 105 198 0 1 168 0 0.0 2 1 2 1 65 1 0 120 177 0 1 140 0 0.4 2 0 3 1 44 1 1 130 219 0 0 188 0 0.0 2 0 2 1 54 1 2 125 273 0 0 152 0 0.5 0 1 2 1 51 1 3 125 213 0 0 125 1 1.4 2 1 2 1 46 0 2 142 177 0 0 160 1 1.4 0 0 2 1 54 0 2 135 304 1 1 170 0 0.0 2 0 2 1 54 1 2 150 232 0 0 165 0 1.6 2 0 3 1 65 0 2 155 269 0 1 148 0 0.8 2 0 2 1 65 0 2 160 360 0 0 151 0 0.8 2 0 2 1 51 0 2 140 308 0 0 142 0 1.5 2 1 2 1 48 1 1 130 245 0 0 180 0 0.2 1 0 2 1 45 1 0 104 208 0 0 148 1 3.0 1 0 2 1 53 0 0 130 264 0 0 143 0 0.4 1 0 2 1 39 1 2 140 321 0 0 182 0 0.0 2 0 2 1 52 1 1 120 325 0 1 172 0 0.2 2 0 2 1 44 1 2 140 235 0 0 180 0 0.0 2 0 2 1 47 1 2 138 257 0 0 156 0 0.0 2 0 2 1 53 0 2 128 216 0 0 115 0 0.0 2 0 0 1 53 0 0 138 234 0 0 160 0 0.0 2 0 2 1 51 0 2 130 256 0 0 149 0 0.5 2 0 2 1 66 1 0 120 302 0 0 151 0 0.4 1 0 2 1 62 1 2 130 231 0 1 146 0 1.8 1 3 3 1 44 0 2 108 141 0 1 175 0 0.6 1 0 2 1 63 0 2 135 252 0 0 172 0 0.0 2 0 2 1 52 1 1 134 201 0 1 158 0 0.8 2 1 2 1 48 1 0 122 222 0 0 186 0 0.0 2 0 2 1 45 1 0 115 260 0 0 185 0 0.0 2 0 2 1 34 1 3 118 182 0 0 174 0 0.0 2 0 2 1 57 0 0 128 303 0 0 159 0 0.0 2 1 2 1 71 0 2 110 265 1 0 130 0 0.0 2 1 2 1 54 1 1 108 309 0 1 156 0 0.0 2 0 3 1 52 1 3 118 186 0 0 190 0 0.0 1 0 1 1 41 1 1 135 203 0 1 132 0 0.0 1 0 1 1 58 1 2 140 211 1 0 165 0 0.0 2 0 2 1 35 0 0 138 183 0 1 182 0 1.4 2 0 2 1 51 1 2 100 222 0 1 143 1 1.2 1 0 2 1 45 0 1 130 234 0 0 175 0 0.6 1 0 2 1 44 1 1 120 220 0 1 170 0 0.0 2 0 2 1 62 0 0 124 209 0 1 163 0 0.0 2 0 2 1 54 1 2 120 258 0 0 147 0 0.4 1 0 3 1 51 1 2 94 227 0 1 154 1 0.0 2 1 3 1 29 1 1 130 204 0 0 202 0 0.0 2 0 2 1 51 1 0 140 261 0 0 186 1 0.0 2 0 2 1 43 0 2 122 213 0 1 165 0 0.2 1 0 2 1 55 0 1 135 250 0 0 161 0 1.4 1 0 2 1 51 1 2 125 245 1 0 166 0 2.4 1 0 2 1 59 1 1 140 221 0 1 164 1 0.0 2 0 2 1 52 1 1 128 205 1 1 184 0 0.0 2 0 2 1 58 1 2 105 240 0 0 154 1 0.6 1 0 3 1 41 1 2 112 250 0 1 179 0 0.0 2 0 2 1 45 1 1 128 308 0 0 170 0 0.0 2 0 2 1 60 0 2 102 318 0 1 160 0 0.0 2 1 2 1 52 1 3 152 298 1 1 178 0 1.2 1 0 3 1 42 0 0 102 265 0 0 122 0 0.6 1 0 2 1 67 0 2 115 564 0 0 160 0 1.6 1 0 3 1 68 1 2 118 277 0 1 151 0 1.0 2 1 3 1 46 1 1 101 197 1 1 156 0 0.0 2 0 3 1 54 0 2 110 214 0 1 158 0 1.6 1 0 2 1 58 0 0 100 248 0 0 122 0 1.0 1 0 2 1 48 1 2 124 255 1 1 175 0 0.0 2 2 2 1 57 1 0 132 207 0 1 168 1 0.0 2 0 3 1 52 1 2 138 223 0 1 169 0 0.0 2 4 2 1 54 0 1 132 288 1 0 159 1 0.0 2 1 2 1 45 0 1 112 160 0 1 138 0 0.0 1 0 2 1 53 1 0 142 226 0 0 111 1 0.0 2 0 3 1 62 0 0 140 394 0 0 157 0 1.2 1 0 2 1 52 1 0 108 233 1 1 147 0 0.1 2 3 3 1 43 1 2 130 315 0 1 162 0 1.9 2 1 2 1 53 1 2 130 246 1 0 173 0 0.0 2 3 2 1 42 1 3 148 244 0 0 178 0 0.8 2 2 2 1 59 1 3 178 270 0 0 145 0 4.2 0 0 3 1 63 0 1 140 195 0 1 179 0 0.0 2 2 2 1 42 1 2 120 240 1 1 194 0 0.8 0 0 3 1 50 1 2 129 196 0 1 163 0 0.0 2 0 2 1 68 0 2 120 211 0 0 115 0 1.5 1 0 2 1

### Columns

In [10]:
### age: The person's age
### sex : 1 = male, 0 = female
### cp : the chest pain experience 
###  trestbps: The person's resting blood pressure
### chol : the person's cholesterol measurement in mg/dl
### fbs: the person's fasting blood sugar (>120mg/dl, 1 = true, 0 =false)
### restecg : resting resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probale or define left ventricular hypertrophy by Estes' criteria)
### thalach : the person's maximum heart rate achieved
### exag: Exercise induced angina ( 1= yes, 0 = no)
### oldpeak: ST depression induced by exercise relative to rest ('ST' related to positions on the ECG plot)
### slope: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3:downsloping)
### ca: The number of major vessels (0-3)
### thal: A blood disorder called thalassemia (3= normal, 6= fixed, 7 =reversable defeat)
### target: Heart disease (0=no, 1= yes)

In [11]:
# With this registered as a temp view, it will only be available to this particular notebook. If you'd like other users to be able to query this table, you can also create a table from the DataFrame.
# Once saved, this table will persist across cluster restarts as well as allow various users across different notebooks to query this data.
# To do so, choose your table name and uncomment the bottom line.

permanent_table_name = "heartdisease_csv"

# df.write.format("parquet").saveAsTable(permanent_table_name)

In [12]:
print("Our dataset has %d rows." % df.count())

Our dataset has 303 rows.

### Since our file is in CSV, we use panda's read_csv to read CSV data file.

In [14]:
import pandas as pd
dt = df.toPandas()

/databricks/spark/python/pyspark/sql/types.py:1636: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
 return pa.schema(fields)

### Check if this is an imbalanced dataset

In [16]:
target_balance = dt['target'].value_counts()

In [17]:
%sql

select target from `HeartDisease_csv` count

target 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0

###Extract DataFrame correlations

In [19]:
dt.corr()

age 
 sex 
 cp 
 trestbps 
 chol 
 fbs 
 restecg 
 thalach 
 exang 
 oldpeak 
 slope 
 ca 
 thal 
 target 
 
 
 
 
 age 
 1.000000 
 -0.098447 
 -0.068653 
 0.279351 
 0.213678 
 0.121308 
 -0.116211 
 -0.398522 
 0.096801 
 0.210013 
 -0.168814 
 0.276326 
 0.068001 
 -0.225439 
 
 
 sex 
 -0.098447 
 1.000000 
 -0.049353 
 -0.056769 
 -0.197912 
 0.045032 
 -0.058196 
 -0.044020 
 0.141664 
 0.096093 
 -0.030711 
 0.118261 
 0.210041 
 -0.280937 
 
 
 cp 
 -0.068653 
 -0.049353 
 1.000000 
 0.047608 
 -0.076904 
 0.094444 
 0.044421 
 0.295762 
 -0.394280 
 -0.149230 
 0.119717 
 -0.181053 
 -0.161736 
 0.433798 
 
 
 trestbps 
 0.279351 
 -0.056769 
 0.047608 
 1.000000 
 0.123174 
 0.177531 
 -0.114103 
 -0.046698 
 0.067616 
 0.193216 
 -0.121475 
 0.101389 
 0.062210 
 -0.144931 
 
 
 chol 
 0.213678 
 -0.197912 
 -0.076904 
 0.123174 
 1.000000 
 0.013294 
 -0.151040 
 -0.009940 
 0.067023 
 0.053952 
 -0.004038 
 0.070511 
 0.098803 
 -0.085239 
 
 
 fbs 
 0.121308 
 0.045032 
 0.094444 
 0.177531 
 0.013294 
 1.000000 
 -0.084189 
 -0.008567 
 0.025665 
 0.005747 
 -0.059894 
 0.137979 
 -0.032019 
 -0.028046 
 
 
 restecg 
 -0.116211 
 -0.058196 
 0.044421 
 -0.114103 
 -0.151040 
 -0.084189 
 1.000000 
 0.044123 
 -0.070733 
 -0.058770 
 0.093045 
 -0.072042 
 -0.011981 
 0.137230 
 
 
 thalach 
 -0.398522 
 -0.044020 
 0.295762 
 -0.046698 
 -0.009940 
 -0.008567 
 0.044123 
 1.000000 
 -0.378812 
 -0.344187 
 0.386784 
 -0.213177 
 -0.096439 
 0.421741 
 
 
 exang 
 0.096801 
 0.141664 
 -0.394280 
 0.067616 
 0.067023 
 0.025665 
 -0.070733 
 -0.378812 
 1.000000 
 0.288223 
 -0.257748 
 0.115739 
 0.206754 
 -0.436757 
 
 
 oldpeak 
 0.210013 
 0.096093 
 -0.149230 
 0.193216 
 0.053952 
 0.005747 
 -0.058770 
 -0.344187 
 0.288223 
 1.000000 
 -0.577537 
 0.222682 
 0.210244 
 -0.430696 
 
 
 slope 
 -0.168814 
 -0.030711 
 0.119717 
 -0.121475 
 -0.004038 
 -0.059894 
 0.093045 
 0.386784 
 -0.257748 
 -0.577537 
 1.000000 
 -0.080155 
 -0.104764 
 0.345877 
 
 
 ca 
 0.276326 
 0.118261 
 -0.181053 
 0.101389 
 0.070511 
 0.137979 
 -0.072042 
 -0.213177 
 0.115739 
 0.222682 
 -0.080155 
 1.000000 
 0.151832 
 -0.391724 
 
 
 thal 
 0.068001 
 0.210041 
 -0.161736 
 0.062210 
 0.098803 
 -0.032019 
 -0.011981 
 -0.096439 
 0.206754 
 0.210244 
 -0.104764 
 0.151832 
 1.000000 
 -0.344029 
 
 
 target 
 -0.225439 
 -0.280937 
 0.433798 
 -0.144931 
 -0.085239 
 -0.028046 
 0.137230 
 0.421741 
 -0.436757 
 -0.430696 
 0.345877 
 -0.391724 
 -0.344029 
 1.000000

### Visualize correlation matrix using seaborn

### High correlation between thal(a blood disorder called thalassemia), oldpeak(ST depression), thalach(the person's maximum heart rate archieved) and target(disease or not diseased)

In [22]:
import seaborn as sns
import matplotlib.pyplot as plt

fig, ax = plt.subplots(figsize=(10,10))
corr = dt.corr()
ax = sns.heatmap(
    corr, 
    vmin=-1, vmax=1, center=0,
    cmap=sns.diverging_palette(50, 220, n=200),
    square=True
)
ax.set_xticklabels(
    ax.get_xticklabels(),
    rotation=45,
    horizontalalignment='right'
)

### The average, minimum and maximum of factors

In [24]:

display(df.describe())


summary age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target count 303 303 303 303 303 303 303 303 303 303 303 303 303 303 mean 54.366336633663366 0.6831683168316832 0.966996699669967 131.62376237623764 246.26402640264027 0.1485148514851485 0.528052805280528 149.64686468646866 0.32673267326732675 1.0396039603960396 1.3993399339933994 0.7293729372937293 2.3135313531353137 0.5445544554455446 stddev 9.08210098983786 0.46601082333962385 1.0320524894832983 17.5381428135171 51.83075098793005 0.35619787492797644 0.525859596359298 22.90516111491409 0.46979446452231655 1.1610750220686346 0.6162261453459622 1.0226063649693276 0.6122765072781408 0.4988347841643915 min 29 0 0 94 126 0 0 71 0 0.0 0 0 0 0 max 77 1 3 200 564 1 2 202 1 6.2 2 4 3 1

### Split the dataset randomly into 70% for training and 30% for testing.

In [26]:
train, test = df.randomSplit([0.7, 0.3], seed = 0)
(train.count(), test.count())
print("We have %d training examples and %d test examples." % (train.count(), test.count()))



We have 223 training examples and 80 test examples.

In [27]:
display(train)

age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target 34 0 1 118 210 0 1 192 0 0.7 2 0 2 1 34 1 3 118 182 0 0 174 0 0.0 2 0 2 1 35 1 1 122 192 0 1 174 0 0.0 2 0 2 1 37 1 2 130 250 0 1 187 0 3.5 0 0 2 1 38 1 2 138 175 0 1 173 0 0.0 2 4 2 1 38 1 2 138 175 0 1 173 0 0.0 2 4 2 1 38 1 3 120 231 0 1 182 1 3.8 1 0 3 0 39 0 2 94 199 0 1 179 0 0.0 2 0 2 1 39 0 2 138 220 0 1 152 0 0.0 1 0 2 1 39 1 0 118 219 0 1 140 0 1.2 1 0 3 0 39 1 2 140 321 0 0 182 0 0.0 2 0 2 1 40 1 0 110 167 0 0 114 1 2.0 1 0 3 0 40 1 0 152 223 0 1 181 0 0.0 2 0 3 0 41 0 1 105 198 0 1 168 0 0.0 2 1 2 1 41 0 1 126 306 0 1 163 0 0.0 2 0 2 1 41 0 1 130 204 0 0 172 0 1.4 2 0 2 1 41 1 1 110 235 0 1 153 0 0.0 2 0 2 1 41 1 1 120 157 0 1 182 0 0.0 2 0 2 1 41 1 2 130 214 0 0 168 0 2.0 1 0 2 1 42 0 0 102 265 0 0 122 0 0.6 1 0 2 1 42 0 2 120 209 0 1 173 0 0.0 1 0 2 1 42 1 0 140 226 0 1 178 0 0.0 2 0 2 1 42 1 1 120 295 0 1 162 0 0.0 2 0 2 1 42 1 2 120 240 1 1 194 0 0.8 0 0 3 1 42 1 2 130 180 0 1 150 0 0.0 2 0 2 1 42 1 3 148 244 0 0 178 0 0.8 2 2 2 1 43 0 0 132 341 1 0 136 1 3.0 1 0 3 0 43 1 0 110 211 0 1 161 0 0.0 2 0 3 1 43 1 0 132 247 1 0 143 1 0.1 1 4 3 0 43 1 0 150 247 0 1 171 0 1.5 2 0 2 1 43 1 2 130 315 0 1 162 0 1.9 2 1 2 1 44 0 2 108 141 0 1 175 0 0.6 1 0 2 1 44 1 0 112 290 0 0 153 0 0.0 2 1 2 0 44 1 0 120 169 0 1 144 1 2.8 0 0 1 0 44 1 1 120 220 0 1 170 0 0.0 2 0 2 1 44 1 1 120 263 0 1 173 0 0.0 2 0 3 1 44 1 2 120 226 0 1 169 0 0.0 2 0 2 1 44 1 2 140 235 0 0 180 0 0.0 2 0 2 1 45 0 0 138 236 0 0 152 1 0.2 1 0 2 1 45 0 1 112 160 0 1 138 0 0.0 1 0 2 1 45 1 0 104 208 0 0 148 1 3.0 1 0 2 1 45 1 0 115 260 0 0 185 0 0.0 2 0 2 1 45 1 1 128 308 0 0 170 0 0.0 2 0 2 1 45 1 3 110 264 0 1 132 0 1.2 1 0 3 0 46 0 2 142 177 0 0 160 1 1.4 0 0 2 1 46 1 1 101 197 1 1 156 0 0.0 2 0 3 1 46 1 2 150 231 0 1 147 0 3.6 1 0 2 0 47 1 0 110 275 0 0 118 1 1.0 1 1 2 0 47 1 2 108 243 0 1 152 0 0.0 2 0 2 0 47 1 2 130 253 0 1 179 0 0.0 2 0 2 1 47 1 2 138 257 0 0 156 0 0.0 2 0 2 1 48 0 2 130 275 0 1 139 0 0.2 2 0 2 1 48 1 0 122 222 0 0 186 0 0.0 2 0 2 1 48 1 0 124 274 0 0 166 0 0.5 1 0 3 0 48 1 0 130 256 1 0 150 1 0.0 2 2 3 0 48 1 1 130 245 0 0 180 0 0.2 1 0 2 1 48 1 2 124 255 1 1 175 0 0.0 2 2 2 1 49 0 0 130 269 0 1 163 0 0.0 2 0 2 1 49 1 2 118 149 0 0 126 0 0.8 2 3 2 0 49 1 2 120 188 0 1 139 0 2.0 1 3 3 0 50 0 0 110 254 0 0 159 0 0.0 2 0 2 1 50 0 1 120 244 0 1 162 0 1.1 2 0 2 1 50 0 2 120 219 0 1 158 0 1.6 1 0 2 1 50 1 0 144 200 0 0 126 1 0.9 1 0 3 0 50 1 2 140 233 0 1 163 0 0.6 1 1 3 0 51 0 0 130 305 0 1 142 1 1.2 1 0 3 0 51 0 2 120 295 0 0 157 0 0.6 2 0 2 1 51 0 2 130 256 0 0 149 0 0.5 2 0 2 1 51 0 2 140 308 0 0 142 0 1.5 2 1 2 1 51 1 0 140 298 0 1 122 1 4.2 1 3 3 0 51 1 0 140 299 0 1 173 1 1.6 2 0 3 0 51 1 2 94 227 0 1 154 1 0.0 2 1 3 1 51 1 2 110 175 0 1 123 0 0.6 2 0 2 1 52 0 2 136 196 0 0 169 0 0.1 1 0 2 1 52 1 0 108 233 1 1 147 0 0.1 2 3 3 1 52 1 0 112 230 0 1 160 0 0.0 2 1 2 0 52 1 0 128 204 1 1 156 1 1.0 1 0 0 0 52 1 0 128 255 0 1 161 1 0.0 2 1 3 0 52 1 1 128 205 1 1 184 0 0.0 2 0 2 1 52 1 1 134 201 0 1 158 0 0.8 2 1 2 1 52 1 2 138 223 0 1 169 0 0.0 2 4 2 1 52 1 2 172 199 1 1 162 0 0.5 2 0 3 1 52 1 3 118 186 0 0 190 0 0.0 1 0 1 1 53 0 0 130 264 0 0 143 0 0.4 1 0 2 1 53 0 2 128 216 0 0 115 0 0.0 2 0 0 1 53 1 0 123 282 0 1 95 1 2.0 1 2 3 0 53 1 0 142 226 0 0 111 1 0.0 2 0 3 1 53 1 2 130 197 1 0 152 0 1.2 0 0 2 1 54 0 1 132 288 1 0 159 1 0.0 2 1 2 1 54 0 2 110 214 0 1 158 0 1.6 1 0 2 1 54 0 2 135 304 1 1 170 0 0.0 2 0 2 1 54 0 2 160 201 0 1 163 0 0.0 2 1 2 1 54 1 0 110 206 0 0 108 1 0.0 1 1 2 0 54 1 0 110 239 0 1 126 1 2.8 1 1 3 0 54 1 0 120 188 0 1 113 0 1.4 1 1 3 0 54 1 0 124 266 0 0 109 1 2.2 1 1 3 0 54 1 1 192 283 0 0 195 0 0.0 2 1 3 0 54 1 2 120 258 0 0 147 0 0.4 1 0 3 1 54 1 2 125 273 0 0 152 0 0.5 0 1 2 1 55 0 0 128 205 0 2 130 1 2.0 1 1 3 0 55 0 0 180 327 0 2 117 1 3.4 1 0 2 0 55 0 1 132 342 0 1 166 0 1.2 2 0 2 1 55 1 0 132 353 0 1 132 1 1.2 1 1 3 0 55 1 0 140 217 0 1 111 1 5.6 0 0 3 0 55 1 0 160 289 0 0 145 1 0.8 1 1 3 0 55 1 1 130 262 0 1 155 0 0.0 2 0 2 1 5

In [28]:
display(test)

age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal target 29 1 1 130 204 0 0 202 0 0.0 2 0 2 1 35 0 0 138 183 0 1 182 0 1.4 2 0 2 1 35 1 0 120 198 0 1 130 1 1.6 1 0 3 0 35 1 0 126 282 0 0 156 1 0.0 2 0 3 0 37 0 2 120 215 0 1 170 0 0.0 2 0 2 1 40 1 3 140 199 0 1 178 1 1.4 2 0 3 1 41 0 2 112 268 0 0 172 1 0.0 2 0 2 1 41 1 0 110 172 0 0 158 0 0.0 2 0 3 0 41 1 1 135 203 0 1 132 0 0.0 1 0 1 1 41 1 2 112 250 0 1 179 0 0.0 2 0 2 1 42 1 0 136 315 0 1 125 1 1.8 1 0 1 0 43 0 2 122 213 0 1 165 0 0.2 1 0 2 1 43 1 0 115 303 0 1 181 0 1.2 1 0 2 1 43 1 0 120 177 0 0 120 1 2.5 1 0 3 0 44 0 2 118 242 0 1 149 0 0.3 1 1 2 1 44 1 0 110 197 0 0 177 0 0.0 2 1 2 0 44 1 1 130 219 0 0 188 0 0.0 2 0 2 1 44 1 2 130 233 0 1 179 1 0.4 2 0 2 1 45 0 1 130 234 0 0 175 0 0.6 1 0 2 1 45 1 0 142 309 0 0 147 1 0.0 1 3 3 0 46 0 0 138 243 0 0 152 1 0.0 1 0 2 1 46 0 1 105 204 0 1 172 0 0.0 2 0 2 1 46 1 0 120 249 0 0 144 0 0.8 2 0 3 0 46 1 0 140 311 0 1 120 1 1.8 1 2 3 0 47 1 0 112 204 0 1 143 0 0.1 2 0 2 1 48 1 1 110 229 0 1 168 0 1.0 0 0 3 0 49 0 1 134 271 0 1 162 0 0.0 1 0 2 1 49 1 1 130 266 0 1 171 0 0.6 2 0 2 1 50 1 0 150 243 0 0 128 0 2.6 1 0 3 0 50 1 2 129 196 0 1 163 0 0.0 2 0 2 1 51 1 0 140 261 0 0 186 1 0.0 2 0 2 1 51 1 2 100 222 0 1 143 1 1.2 1 0 2 1 51 1 2 125 245 1 0 166 0 2.4 1 0 2 1 51 1 3 125 213 0 0 125 1 1.4 2 1 2 1 52 1 0 125 212 0 1 168 0 1.0 2 2 3 0 52 1 1 120 325 0 1 172 0 0.2 2 0 2 1 52 1 3 152 298 1 1 178 0 1.2 1 0 3 1 53 0 0 138 234 0 0 160 0 0.0 2 0 2 1 53 1 0 140 203 1 0 155 1 3.1 0 0 3 0 53 1 2 130 246 1 0 173 0 0.0 2 3 2 1 54 0 2 108 267 0 0 167 0 0.0 2 0 2 1 54 1 0 122 286 0 0 116 1 3.2 1 2 2 0 54 1 0 140 239 0 1 160 0 1.2 2 0 2 1 54 1 1 108 309 0 1 156 0 0.0 2 0 3 1 54 1 2 150 232 0 0 165 0 1.6 2 0 3 1 55 0 1 135 250 0 0 161 0 1.4 1 0 2 1 56 0 0 200 288 1 0 133 1 4.0 0 2 3 0 56 1 1 120 236 0 1 178 0 0.8 2 0 2 1 56 1 1 130 221 0 0 163 0 0.0 2 0 3 1 57 0 1 130 236 0 0 174 0 0.0 1 1 2 0 57 1 0 110 201 0 1 126 1 1.5 1 0 1 1 57 1 0 132 207 0 1 168 1 0.0 2 0 3 1 57 1 0 140 192 0 1 148 0 0.4 1 0 1 1 57 1 0 165 289 1 0 124 0 1.0 1 3 3 0 57 1 1 154 232 0 0 164 0 0.0 2 1 2 0 57 1 2 150 168 0 1 174 0 1.6 2 0 2 1 58 1 0 150 270 0 0 111 1 0.8 2 0 3 0 58 1 1 125 220 0 1 144 0 0.4 1 4 3 1 58 1 2 105 240 0 0 154 1 0.6 1 0 3 1 58 1 2 132 224 0 0 173 0 3.2 2 2 3 0 58 1 2 140 211 1 0 165 0 0.0 2 0 2 1 59 1 3 134 204 0 1 162 0 0.8 2 2 2 0 60 0 2 102 318 0 1 160 0 0.0 2 1 2 1 60 1 0 125 258 0 0 141 1 2.8 1 1 3 0 61 1 0 138 166 0 0 125 1 3.6 1 1 2 0 62 0 0 160 164 0 0 145 0 6.2 0 3 3 0 62 1 2 130 231 0 1 146 0 1.8 1 3 3 1 63 0 0 124 197 0 1 136 1 0.0 1 0 2 0 63 0 0 150 407 0 0 154 0 4.0 1 3 3 0 63 0 2 135 252 0 0 172 0 0.0 2 0 2 1 64 1 2 140 335 0 1 158 0 0.0 2 0 2 0 65 0 0 150 225 0 0 114 0 1.0 1 3 3 0 65 0 2 140 417 1 0 157 0 0.8 2 1 2 1 65 1 0 135 254 0 0 127 0 2.8 1 1 3 0 66 1 1 160 246 0 1 120 1 0.0 1 3 1 0 67 0 2 152 277 0 1 172 0 0.0 2 1 2 1 67 1 0 120 237 0 1 71 0 1.0 1 0 2 0 68 0 2 120 211 0 0 115 0 1.5 1 0 2 1 68 1 2 118 277 0 1 151 0 1.0 2 1 3 1 74 0 1 120 269 0 0 121 1 0.2 2 1 2 1

### Data visualization

###  Older people are more likely than younger people to suffer from Heart disease.

In [31]:
display(train.select("target", "age"))

target age 1 34 1 34 1 35 1 37 1 38 1 38 0 38 1 39 1 39 0 39 1 39 0 40 0 40 1 41 1 41 1 41 1 41 1 41 1 41 1 42 1 42 1 42 1 42 1 42 1 42 1 42 0 43 1 43 0 43 1 43 1 43 1 44 0 44 0 44 1 44 1 44 1 44 1 44 1 45 1 45 1 45 1 45 1 45 0 45 1 46 1 46 0 46 0 47 0 47 1 47 1 47 1 48 1 48 0 48 0 48 1 48 1 48 1 49 0 49 0 49 1 50 1 50 1 50 0 50 0 50 0 51 1 51 1 51 1 51 0 51 0 51 1 51 1 51 1 52 1 52 0 52 0 52 0 52 1 52 1 52 1 52 1 52 1 52 1 53 1 53 0 53 1 53 1 53 1 54 1 54 1 54 1 54 0 54 0 54 0 54 0 54 0 54 1 54 1 54 0 55 0 55 1 55 0 55 0 55 0 55 1 55 0 56 1 56 0 56 0 56 0 56 1 56 0 56 1 56 1 57 1 57 0 57 0 57 0 57 0 57 0 57 0 57 0 57 1 57 1 58 1 58 0 58 0 58 1 58 1 58 0 58 0 58 0 58 0 58 0 58 0 58 0 58 0 58 0 59 0 59 1 59 1 59 0 59 0 59 0 59 1 59 0 59 1 59 0 59 0 59 1 59 0 60 1 60 1 60 0 60 0 60 0 60 0 60 0 60 0 60 0 61 0 61 0 61 0 61 0 61 1 61 0 61 1 62 0 62 0 62 1 62 0 62 0 62 0 62 0 62 1 62 0 63 1 63 0 63 0 63 0 63 1 63 1 64 1 64 1 64 0 64 1 64 0 64 0 64 1 64 1 64 1 65 1 65 0 65 1 65 0 65 0 66 1 66 1 66 0 66 1 66 1 66 1 67 1 67 0 67 0 67 0 67 0 67 0 67 0 68 0 68 1 69 0 69 1 69 0 70 0 70 1 70 0 70 1 71 1 71 1 71 1 76 0 77

In [32]:
display(train.select("sex", "target"))
#sex: The person's sex (1 = male, 0 = female)

sex target 0 1 1 1 1 1 1 1 1 1 1 1 1 0 0 1 0 1 1 0 1 1 1 0 1 0 0 1 0 1 0 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 1 1 1 0 0 1 1 1 0 1 1 1 1 0 1 1 0 1 0 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 0 1 0 1 0 1 1 1 1 0 1 1 1 1 0 1 0 1 1 1 1 0 1 1 0 1 0 0 1 0 1 0 1 1 0 1 0 0 0 0 1 0 1 0 1 1 0 1 0 1 1 1 1 0 1 1 1 1 0 1 0 1 0 1 1 1 1 1 1 1 1 1 1 0 1 0 1 1 0 1 1 1 1 0 1 0 1 0 1 0 1 1 0 1 0 1 0 1 0 1 0 1 1 1 1 0 0 0 0 0 1 1 0 1 0 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 1 0 1 1 0 1 0 1 0 0 1 0 1 0 1 0 1 0 1 0 1 0 1 1 0 1 0 1 0 0 0 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 0 0 1 0 1 1 1 1 1 0 1 0 1 0 1 1 1 0 1 1 1 0 1 0 1 1 0 0 0 1 0 1 1 0 1 0 1 0 1 0 1 0 1 0 0 0 0 0 1 0 1 0 1 0 1 1 1 0 0 1 0 0 0 0 0 1 0 0 0 0 1 0 1 0 1 1 0 0 0 1 1 0 1 0 1 0 1 1 0 1 0 1 0 1 1 0 1 1 1 0 1 0 1 1 1 1 0 1 0 1 1 0 1 1 1 0 0 0 0 1 0 1 1 0 1 1 1 1 0 1 0 1 1 0 1 0 1 0 1 0 1 0 1 0 1 0 0 1 1 0 1 1 1 0 1 0 1 1 1 0 0 1 0 1 0 1 0 1 1 0

### High blood pressure is a risk factor for heart condition

In [34]:
display(train.select("target", "trestbps"))

target trestbps 1 118 1 118 1 122 1 130 1 138 1 138 0 120 1 94 1 138 0 118 1 140 0 110 0 152 1 105 1 126 1 130 1 110 1 120 1 130 1 102 1 120 1 140 1 120 1 120 1 130 1 148 0 132 1 110 0 132 1 150 1 130 1 108 0 112 0 120 1 120 1 120 1 120 1 140 1 138 1 112 1 104 1 115 1 128 0 110 1 142 1 101 0 150 0 110 0 108 1 130 1 138 1 130 1 122 0 124 0 130 1 130 1 124 1 130 0 118 0 120 1 110 1 120 1 120 0 144 0 140 0 130 1 120 1 130 1 140 0 140 0 140 1 94 1 110 1 136 1 108 0 112 0 128 0 128 1 128 1 134 1 138 1 172 1 118 1 130 1 128 0 123 1 142 1 130 1 132 1 110 1 135 1 160 0 110 0 110 0 120 0 124 0 192 1 120 1 125 0 128 0 180 1 132 0 132 0 140 0 160 1 130 0 134 1 140 0 125 0 130 0 132 1 120 0 130 1 120 1 120 1 128 0 140 0 110 0 130 0 150 0 152 0 124 0 128 1 150 1 100 1 130 0 170 0 136 1 120 1 150 0 100 0 114 0 125 0 128 0 128 0 146 0 120 0 112 0 174 0 110 1 135 1 138 0 140 0 164 0 170 1 140 0 126 1 150 0 160 0 170 1 178 0 150 1 120 1 150 0 117 0 130 0 130 0 140 0 145 0 140 0 130 0 145 0 120 0 140 0 148 1 150 0 134 1 124 0 138 0 140 1 140 0 150 0 130 0 120 0 120 1 128 0 108 1 140 0 130 0 130 0 140 1 145 1 130 1 180 1 140 0 120 1 128 0 145 0 125 1 110 1 170 1 155 1 160 0 110 1 120 0 138 0 178 1 146 1 150 0 112 1 120 1 160 1 106 1 115 0 100 0 120 0 125 0 160 0 152 0 144 0 180 1 140 0 140 1 160 0 130 0 145 1 156 0 160 1 112 1 160 1 110 1 140 0 125

### Going higher than your maximum heart rate for long periods of time could be a risk factor for heart condition

In [36]:
display(train.select("target", "thalach"))

target thalach 1 192 1 174 1 174 1 187 1 173 1 173 0 182 1 179 1 152 0 140 1 182 0 114 0 181 1 168 1 163 1 172 1 153 1 182 1 168 1 122 1 173 1 178 1 162 1 194 1 150 1 178 0 136 1 161 0 143 1 171 1 162 1 175 0 153 0 144 1 170 1 173 1 169 1 180 1 152 1 138 1 148 1 185 1 170 0 132 1 160 1 156 0 147 0 118 0 152 1 179 1 156 1 139 1 186 0 166 0 150 1 180 1 175 1 163 0 126 0 139 1 159 1 162 1 158 0 126 0 163 0 142 1 157 1 149 1 142 0 122 0 173 1 154 1 123 1 169 1 147 0 160 0 156 0 161 1 184 1 158 1 169 1 162 1 190 1 143 1 115 0 95 1 111 1 152 1 159 1 158 1 170 1 163 0 108 0 126 0 113 0 109 0 195 1 147 1 152 0 130 0 117 1 166 0 132 0 111 0 145 1 155 0 150 1 153 0 144 0 103 0 105 1 169 0 142 1 162 1 163 1 159 0 123 0 143 0 115 0 112 0 88 0 141 0 150 1 173 1 122 1 131 0 146 0 152 1 172 1 162 0 156 0 140 0 171 0 131 0 130 0 105 0 160 0 165 0 143 0 142 1 161 1 182 0 162 0 90 0 140 1 164 0 134 1 157 0 125 0 159 1 145 0 157 1 96 1 171 0 160 0 132 0 144 0 170 0 142 0 155 0 169 0 146 0 140 0 138 0 161 1 137 0 145 1 163 0 106 0 160 1 157 0 154 0 97 0 99 0 103 1 140 0 169 1 179 0 147 0 132 0 144 1 150 1 122 1 154 1 133 0 96 1 105 0 132 0 131 1 144 1 155 1 148 1 151 0 158 1 140 0 174 0 165 1 152 1 114 0 132 1 151 1 138 1 142 1 160 0 125 0 129 0 163 0 108 0 150 0 141 0 150 1 151 0 146 1 131 0 109 0 125 1 143 0 112 1 125 1 162 1 130 1 116 0 162

### When there is high cholesterol in your blood, it builds up in the walls of your arteries, causing a form of heart disease.

In [38]:
display(train.select("target", "chol"))

target chol 1 210 1 182 1 192 1 250 1 175 1 175 0 231 1 199 1 220 0 219 1 321 0 167 0 223 1 198 1 306 1 204 1 235 1 157 1 214 1 265 1 209 1 226 1 295 1 240 1 180 1 244 0 341 1 211 0 247 1 247 1 315 1 141 0 290 0 169 1 220 1 263 1 226 1 235 1 236 1 160 1 208 1 260 1 308 0 264 1 177 1 197 0 231 0 275 0 243 1 253 1 257 1 275 1 222 0 274 0 256 1 245 1 255 1 269 0 149 0 188 1 254 1 244 1 219 0 200 0 233 0 305 1 295 1 256 1 308 0 298 0 299 1 227 1 175 1 196 1 233 0 230 0 204 0 255 1 205 1 201 1 223 1 199 1 186 1 264 1 216 0 282 1 226 1 197 1 288 1 214 1 304 1 201 0 206 0 239 0 188 0 266 0 283 1 258 1 273 0 205 0 327 1 342 0 353 0 217 0 289 1 262 0 409 1 294 0 249 0 283 0 184 1 240 0 256 1 193 1 354 1 303 0 241 0 335 0 131 0 276 0 274 0 261 0 229 1 126 1 248 1 197 0 225 0 319 1 340 1 283 0 234 0 318 0 300 0 216 0 259 0 218 0 284 0 230 0 249 0 239 1 234 1 271 0 177 0 176 0 326 1 221 0 218 1 212 0 273 0 288 1 270 0 258 1 178 1 240 0 230 0 206 0 253 0 293 0 282 0 185 0 330 0 307 0 260 0 207 0 203 1 243 0 234 1 209 0 294 0 268 1 394 0 244 0 263 0 267 0 281 1 208 0 269 1 195 0 254 0 330 0 187 1 233 1 303 1 325 1 313 0 246 1 263 0 212 0 309 1 211 1 227 1 269 1 360 0 248 1 177 0 282 0 228 1 278 1 226 0 212 1 302 1 228 1 223 1 564 0 299 0 229 0 254 0 286 0 212 0 193 0 274 1 239 0 254 1 234 0 322 0 174 1 245 0 269 1 149 1 302 1 265 1 197 0 304

###Deeper and more widespread ST depression generally indicates more severe or extensive disease.

In [40]:
display(train.select("target","oldpeak"))

target oldpeak 1 0.7 1 0.0 1 0.0 1 3.5 1 0.0 1 0.0 0 3.8 1 0.0 1 0.0 0 1.2 1 0.0 0 2.0 0 0.0 1 0.0 1 0.0 1 1.4 1 0.0 1 0.0 1 2.0 1 0.6 1 0.0 1 0.0 1 0.0 1 0.8 1 0.0 1 0.8 0 3.0 1 0.0 0 0.1 1 1.5 1 1.9 1 0.6 0 0.0 0 2.8 1 0.0 1 0.0 1 0.0 1 0.0 1 0.2 1 0.0 1 3.0 1 0.0 1 0.0 0 1.2 1 1.4 1 0.0 0 3.6 0 1.0 0 0.0 1 0.0 1 0.0 1 0.2 1 0.0 0 0.5 0 0.0 1 0.2 1 0.0 1 0.0 0 0.8 0 2.0 1 0.0 1 1.1 1 1.6 0 0.9 0 0.6 0 1.2 1 0.6 1 0.5 1 1.5 0 4.2 0 1.6 1 0.0 1 0.6 1 0.1 1 0.1 0 0.0 0 1.0 0 0.0 1 0.0 1 0.8 1 0.0 1 0.5 1 0.0 1 0.4 1 0.0 0 2.0 1 0.0 1 1.2 1 0.0 1 1.6 1 0.0 1 0.0 0 0.0 0 2.8 0 1.4 0 2.2 0 0.0 1 0.4 1 0.5 0 2.0 0 3.4 1 1.2 0 1.2 0 5.6 0 0.8 1 0.0 0 1.9 1 1.3 0 1.2 0 1.6 0 2.1 1 0.0 0 0.6 1 1.9 1 0.6 1 0.0 0 0.2 0 3.0 0 1.2 0 0.6 0 1.2 0 0.3 0 0.4 1 0.2 1 1.0 1 0.6 0 2.8 0 0.0 1 0.0 1 1.0 0 0.1 0 4.4 0 0.0 0 2.2 0 3.0 0 2.0 0 1.8 0 2.5 0 0.0 0 1.2 1 0.5 1 0.0 0 0.0 0 1.0 0 3.4 1 0.0 0 2.2 1 1.6 0 0.0 0 0.2 1 4.2 0 2.6 1 0.0 1 0.9 0 1.4 0 2.4 0 1.4 0 1.2 0 2.8 0 3.0 0 0.0 0 1.0 0 3.6 0 1.9 0 0.0 1 1.0 0 2.6 1 0.0 0 1.9 0 3.6 1 1.2 0 1.4 0 1.2 0 1.8 0 1.4 1 0.0 0 1.8 1 0.0 0 1.4 0 1.8 0 4.0 1 2.3 1 2.0 1 0.0 1 0.2 0 2.2 1 0.2 0 2.0 0 1.8 1 1.8 1 0.6 1 0.8 1 0.8 0 0.6 1 0.4 0 1.4 0 1.0 1 0.0 1 2.6 0 0.1 1 0.4 1 2.3 1 0.3 1 1.6 0 0.9 0 2.6 0 0.2 0 1.5 0 0.8 0 3.4 0 1.6 1 1.8 0 2.0 1 0.1 0 2.4 0 2.6 1 0.0 0 2.9 1 1.6 1 0.4 1 0.0 1 1.1 0 0.0

### Thalassemia(thal) is a blood disorder that causes heart disease

In [42]:
display(train.select("target","thal"))

target thal 1 2 1 2 1 2 1 2 1 2 1 2 0 3 1 2 1 2 0 3 1 2 0 3 0 3 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 2 1 3 1 2 1 2 0 3 1 3 0 3 1 2 1 2 1 2 0 2 0 1 1 2 1 3 1 2 1 2 1 2 1 2 1 2 1 2 1 2 0 3 1 2 1 3 0 2 0 2 0 2 1 2 1 2 1 2 1 2 0 3 0 3 1 2 1 2 1 2 0 2 0 3 1 2 1 2 1 2 0 3 0 3 0 3 1 2 1 2 1 2 0 3 0 3 1 3 1 2 1 2 1 3 0 2 0 0 0 3 1 2 1 2 1 2 1 3 1 1 1 2 1 0 0 3 1 3 1 2 1 2 1 2 1 2 1 2 0 2 0 3 0 3 0 3 0 3 1 3 1 2 0 3 0 2 1 2 0 3 0 3 0 3 1 2 0 3 1 2 0 2 0 3 0 1 1 2 0 1 1 3 1 2 1 2 0 3 0 3 0 3 0 1 0 3 0 3 0 3 1 3 1 2 1 2 0 1 0 2 1 2 1 2 0 3 0 1 0 3 0 3 0 3 0 3 0 2 0 3 0 2 0 3 1 3 1 2 0 3 0 1 0 3 1 2 0 1 1 2 0 2 0 3 1 3 0 3 1 2 1 2 0 3 0 3 0 3 0 3 0 3 0 2 0 2 0 3 0 3 0 3 0 3 1 2 0 2 1 2 0 2 0 2 1 2 0 2 0 3 0 3 0 3 1 2 0 2 1 2 0 3 0 3 0 3 1 1 1 2 1 2 1 3 0 2 1 3 0 1 0 3 1 2 1 3 1 2 1 2 0 1 1 3 0 2 0 3 1 2 1 2 0 2 1 2 1 1 1 2 1 3 0 2 0 3 0 3 0 2 0 3 0 3 0 3 1 2 0 3 1 2 0 2 0 3 1 2 0 3 1 2 1 2 1 2 1 2 0 2

In [43]:
from pyspark.ml.feature import VectorAssembler, VectorIndexer
featuresCols = df.columns
featuresCols.remove('target')
# This concatenates all feature columns into a single feature vector in a new column "rawFeatures".
vectorAssembler = VectorAssembler(inputCols=featuresCols, outputCol="rawFeatures")
# This identifies categorical features and indexes them.
vectorIndexer = VectorIndexer(inputCol="rawFeatures", outputCol="features", maxCategories=4)

In [44]:
from pyspark.ml.classification import GBTClassifier, DecisionTreeClassifier
# ml.classification import decisiontree
# Takes the "features" column and learns to predict "target"
gbt = GBTClassifier(labelCol="target", maxDepth=3)
#dt = DecisionTreeClassifier(labelCol="target", featuresCol="features", maxDepth=3)setLabelCol()
#dt = DecisionTreeClassifier()

In [45]:
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
from pyspark.ml.evaluation import BinaryClassificationEvaluator
# Define a grid of hyperparameters to test:
#  - maxDepth: max depth of each decision tree in the GBT ensemble
#  - maxIter: iterations, i.e., number of trees in each GBT ensemble
# In this example notebook, we keep these values small.  In practice, to get the highest accuracy, you would likely want to try deeper trees (10 or higher) and more trees in the ensemble (>100).
paramGrid = ParamGridBuilder()\
  .addGrid(gbt.maxDepth, [2])\
  .addGrid(gbt.maxIter, [10])\
  .build()
# We define an evaluation metric.  This tells CrossValidator how well we are doing by comparing the true labels with predictions.
#evaluator = RegressionEvaluator(metricName="rmse", labelCol=gbt.getLabelCol(), predictionCol=gbt.getPredictionCol())
evaluator = BinaryClassificationEvaluator(metricName="areaUnderROC", labelCol=gbt.getLabelCol())
#evaluator = BinaryClassificationEvaluator()
# Declare the CrossValidator, which runs model tuning for us.
cv = CrossValidator(estimator=gbt, evaluator=evaluator, estimatorParamMaps=paramGrid, numFolds=5)

In [46]:
from pyspark.ml import Pipeline
pipeline = Pipeline(stages=[vectorAssembler, vectorIndexer, cv])


In [47]:
pipelineModel = pipeline.fit(train)

MLlib will automatically track trials in MLflow. After your tuning fit() call has completed, view the MLflow UI to see logged runs.

In [48]:
predictions = pipelineModel.transform(test)

In [49]:
display(predictions.select("target", "prediction", *featuresCols))

target prediction age sex cp trestbps chol fbs restecg thalach exang oldpeak slope ca thal 1 1.0 29 1 1 130 204 0 0 202 0 0.0 2 0 2 1 1.0 35 0 0 138 183 0 1 182 0 1.4 2 0 2 0 0.0 35 1 0 120 198 0 1 130 1 1.6 1 0 3 0 0.0 35 1 0 126 282 0 0 156 1 0.0 2 0 3 1 1.0 37 0 2 120 215 0 1 170 0 0.0 2 0 2 1 1.0 40 1 3 140 199 0 1 178 1 1.4 2 0 3 1 1.0 41 0 2 112 268 0 0 172 1 0.0 2 0 2 0 1.0 41 1 0 110 172 0 0 158 0 0.0 2 0 3 1 1.0 41 1 1 135 203 0 1 132 0 0.0 1 0 1 1 1.0 41 1 2 112 250 0 1 179 0 0.0 2 0 2 0 0.0 42 1 0 136 315 0 1 125 1 1.8 1 0 1 1 1.0 43 0 2 122 213 0 1 165 0 0.2 1 0 2 1 1.0 43 1 0 115 303 0 1 181 0 1.2 1 0 2 0 0.0 43 1 0 120 177 0 0 120 1 2.5 1 0 3 1 1.0 44 0 2 118 242 0 1 149 0 0.3 1 1 2 0 1.0 44 1 0 110 197 0 0 177 0 0.0 2 1 2 1 1.0 44 1 1 130 219 0 0 188 0 0.0 2 0 2 1 1.0 44 1 2 130 233 0 1 179 1 0.4 2 0 2 1 1.0 45 0 1 130 234 0 0 175 0 0.6 1 0 2 0 0.0 45 1 0 142 309 0 0 147 1 0.0 1 3 3 1 0.0 46 0 0 138 243 0 0 152 1 0.0 1 0 2 1 1.0 46 0 1 105 204 0 1 172 0 0.0 2 0 2 0 0.0 46 1 0 120 249 0 0 144 0 0.8 2 0 3 0 0.0 46 1 0 140 311 0 1 120 1 1.8 1 2 3 1 1.0 47 1 0 112 204 0 1 143 0 0.1 2 0 2 0 1.0 48 1 1 110 229 0 1 168 0 1.0 0 0 3 1 1.0 49 0 1 134 271 0 1 162 0 0.0 1 0 2 1 1.0 49 1 1 130 266 0 1 171 0 0.6 2 0 2 0 0.0 50 1 0 150 243 0 0 128 0 2.6 1 0 3 1 1.0 50 1 2 129 196 0 1 163 0 0.0 2 0 2 1 0.0 51 1 0 140 261 0 0 186 1 0.0 2 0 2 1 0.0 51 1 2 100 222 0 1 143 1 1.2 1 0 2 1 1.0 51 1 2 125 245 1 0 166 0 2.4 1 0 2 1 0.0 51 1 3 125 213 0 0 125 1 1.4 2 1 2 0 0.0 52 1 0 125 212 0 1 168 0 1.0 2 2 3 1 1.0 52 1 1 120 325 0 1 172 0 0.2 2 0 2 1 0.0 52 1 3 152 298 1 1 178 0 1.2 1 0 3 1 1.0 53 0 0 138 234 0 0 160 0 0.0 2 0 2 0 0.0 53 1 0 140 203 1 0 155 1 3.1 0 0 3 1 1.0 53 1 2 130 246 1 0 173 0 0.0 2 3 2 1 1.0 54 0 2 108 267 0 0 167 0 0.0 2 0 2 0 0.0 54 1 0 122 286 0 0 116 1 3.2 1 2 2 1 1.0 54 1 0 140 239 0 1 160 0 1.2 2 0 2 1 1.0 54 1 1 108 309 0 1 156 0 0.0 2 0 3 1 1.0 54 1 2 150 232 0 0 165 0 1.6 2 0 3 1 1.0 55 0 1 135 250 0 0 161 0 1.4 1 0 2 0 0.0 56 0 0 200 288 1 0 133 1 4.0 0 2 3 1 1.0 56 1 1 120 236 0 1 178 0 0.8 2 0 2 1 1.0 56 1 1 130 221 0 0 163 0 0.0 2 0 3 0 1.0 57 0 1 130 236 0 0 174 0 0.0 1 1 2 1 0.0 57 1 0 110 201 0 1 126 1 1.5 1 0 1 1 1.0 57 1 0 132 207 0 1 168 1 0.0 2 0 3 1 1.0 57 1 0 140 192 0 1 148 0 0.4 1 0 1 0 0.0 57 1 0 165 289 1 0 124 0 1.0 1 3 3 0 1.0 57 1 1 154 232 0 0 164 0 0.0 2 1 2 1 1.0 57 1 2 150 168 0 1 174 0 1.6 2 0 2 0 0.0 58 1 0 150 270 0 0 111 1 0.8 2 0 3 1 0.0 58 1 1 125 220 0 1 144 0 0.4 1 4 3 1 1.0 58 1 2 105 240 0 0 154 1 0.6 1 0 3 0 0.0 58 1 2 132 224 0 0 173 0 3.2 2 2 3 1 1.0 58 1 2 140 211 1 0 165 0 0.0 2 0 2 0 1.0 59 1 3 134 204 0 1 162 0 0.8 2 2 2 1 1.0 60 0 2 102 318 0 1 160 0 0.0 2 1 2 0 0.0 60 1 0 125 258 0 0 141 1 2.8 1 1 3 0 0.0 61 1 0 138 166 0 0 125 1 3.6 1 1 2 0 0.0 62 0 0 160 164 0 0 145 0 6.2 0 3 3 1 0.0 62 1 2 130 231 0 1 146 0 1.8 1 3 3 0 1.0 63 0 0 124 197 0 1 136 1 0.0 1 0 2 0 0.0 63 0 0 150 407 0 0 154 0 4.0 1 3 3 1 1.0 63 0 2 135 252 0 0 172 0 0.0 2 0 2 0 1.0 64 1 2 140 335 0 1 158 0 0.0 2 0 2 0 0.0 65 0 0 150 225 0 0 114 0 1.0 1 3 3 1 1.0 65 0 2 140 417 1 0 157 0 0.8 2 1 2 0 0.0 65 1 0 135 254 0 0 127 0 2.8 1 1 3 0 0.0 66 1 1 160 246 0 1 120 1 0.0 1 3 1 1 1.0 67 0 2 152 277 0 1 172 0 0.0 2 1 2 0 1.0 67 1 0 120 237 0 1 71 0 1.0 1 0 2 1 1.0 68 0 2 120 211 0 0 115 0 1.5 1 0 2 1 0.0 68 1 2 118 277 0 1 151 0 1.0 2 1 3 1 0.0 74 0 1 120 269 0 0 121 1 0.2 2 1 2

###Training and Making Predictions from sklearn

### The x contains all the columsn from the dataset except the 'target' Columns. The y varaible contains the value from the 'target' Columns

In [52]:
# Separating features(X) and target(y)
X = dt.drop('target', axis=1)

In [53]:
print(X)

age sex cp trestbps chol ... exang oldpeak slope ca thal
0 63 1 3 145 233 ... 0 2.3 0 0 1
1 37 1 2 130 250 ... 0 3.5 0 0 2
2 41 0 1 130 204 ... 0 1.4 2 0 2
3 56 1 1 120 236 ... 0 0.8 2 0 2
4 57 0 0 120 354 ... 1 0.6 2 0 2
5 57 1 0 140 192 ... 0 0.4 1 0 1
6 56 0 1 140 294 ... 0 1.3 1 0 2
7 44 1 1 120 263 ... 0 0.0 2 0 3
8 52 1 2 172 199 ... 0 0.5 2 0 3
9 57 1 2 150 168 ... 0 1.6 2 0 2
10 54 1 0 140 239 ... 0 1.2 2 0 2
11 48 0 2 130 275 ... 0 0.2 2 0 2
12 49 1 1 130 266 ... 0 0.6 2 0 2
13 64 1 3 110 211 ... 1 1.8 1 0 2
14 58 0 3 150 283 ... 0 1.0 2 0 2
15 50 0 2 120 219 ... 0 1.6 1 0 2
16 58 0 2 120 340 ... 0 0.0 2 0 2
17 66 0 3 150 226 ... 0 2.6 0 0 2
18 43 1 0 150 247 ... 0 1.5 2 0 2
19 69 0 3 140 239 ... 0 1.8 2 2 2
20 59 1 0 135 234 ... 0 0.5 1 0 3
21 44 1 2 130 233 ... 1 0.4 2 0 2
22 42 1 0 140 226 ... 0 0.0 2 0 2
23 61 1 2 150 243 ... 1 1.0 1 0 2
24 40 1 3 140 199 ... 1 1.4 2 0 3
25 71 0 1 160 302 ... 0 0.4 2 2 2
26 59 1 2 150 212 ... 0 1.6 2 0 2
27 51 1 2 110 175 ... 0 0.6 2 0 2
28 65 0 2 140 417 ... 0 0.8 2 1 2
29 53 1 2 130 197 ... 0 1.2 0 0 2
.. ... ... .. ... ... ... ... ... ... .. ...
273 58 1 0 100 234 ... 0 0.1 2 1 3
274 47 1 0 110 275 ... 1 1.0 1 1 2
275 52 1 0 125 212 ... 0 1.0 2 2 3
276 58 1 0 146 218 ... 0 2.0 1 1 3
277 57 1 1 124 261 ... 0 0.3 2 0 3
278 58 0 1 136 319 ... 0 0.0 2 2 2
279 61 1 0 138 166 ... 1 3.6 1 1 2
280 42 1 0 136 315 ... 1 1.8 1 0 1
281 52 1 0 128 204 ... 1 1.0 1 0 0
282 59 1 2 126 218 ... 0 2.2 1 1 1
283 40 1 0 152 223 ... 0 0.0 2 0 3
284 61 1 0 140 207 ... 1 1.9 2 1 3
285 46 1 0 140 311 ... 1 1.8 1 2 3
286 59 1 3 134 204 ... 0 0.8 2 2 2
287 57 1 1 154 232 ... 0 0.0 2 1 2
288 57 1 0 110 335 ... 1 3.0 1 1 3
289 55 0 0 128 205 ... 1 2.0 1 1 3
290 61 1 0 148 203 ... 0 0.0 2 1 3
291 58 1 0 114 318 ... 0 4.4 0 3 1
292 58 0 0 170 225 ... 1 2.8 1 2 1
293 67 1 2 152 212 ... 0 0.8 1 0 3
294 44 1 0 120 169 ... 1 2.8 0 0 1
295 63 1 0 140 187 ... 1 4.0 2 2 3
296 63 0 0 124 197 ... 1 0.0 1 0 2
297 59 1 0 164 176 ... 0 1.0 1 2 1
298 57 0 0 140 241 ... 1 0.2 1 0 3
299 45 1 3 110 264 ... 0 1.2 1 0 3
300 68 1 0 144 193 ... 0 3.4 1 2 3
301 57 1 0 130 131 ... 1 1.2 1 1 3
302 57 0 1 130 236 ... 0 0.0 1 1 2

[303 rows x 13 columns]

In [54]:
y = dt['target']
print(y)

0 1
1 1
2 1
3 1
4 1
5 1
6 1
7 1
8 1
9 1
10 1
11 1
12 1
13 1
14 1
15 1
16 1
17 1
18 1
19 1
20 1
21 1
22 1
23 1
24 1
25 1
26 1
27 1
28 1
29 1
 ..
273 0
274 0
275 0
276 0
277 0
278 0
279 0
280 0
281 0
282 0
283 0
284 0
285 0
286 0
287 0
288 0
289 0
290 0
291 0
292 0
293 0
294 0
295 0
296 0
297 0
298 0
299 0
300 0
301 0
302 0
Name: target, Length: 303, dtype: int32

### We use split up 35% of the data in to the test set and 65% for training

In [56]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35)

In [57]:
print(f"X.shape: {X.shape}, y.shape: {y.shape}")
#original dataset

X.shape: (303, 13), y.shape: (303,)

In [58]:
print(f"X_train.shape: {X_train.shape}, y_train.shape: {y_train.shape}")
print(f"X_test.shape: {X_test.shape}, y_test.shape: {y_test.shape}")
#splited datasets

X_train.shape: (196, 13), y_train.shape: (196,)
X_test.shape: (107, 13), y_test.shape: (107,)

### Use the DecisionTreeClassifier to train the algorithm

In [60]:
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier()
classifier.fit(X_train, y_train)

Out[33]: DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
 max_features=None, max_leaf_nodes=None,
 min_impurity_decrease=0.0, min_impurity_split=None,
 min_samples_leaf=1, min_samples_split=2,
 min_weight_fraction_leaf=0.0, presort=False, random_state=None,
 splitter='best')

In [61]:
y_pred = classifier.predict(X_test)
print(y_pred)

[1 0 0 0 0 0 1 0 0 1 1 1 0 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 0 0 1 0 0 1 1 1 0
 0 0 0 0 0 0 1 1 1 1 0 1 1 0 1 0 0 1 1 1 0 0 1 0 1 1 1 1 0 1 1 0 1 1 1 1 0
 1 0 1 0 1 1 1 1 1 1 0 1 0 1 1 1 0 1 0 0 1 1 0 0 0 0 1 1 1 1 1 0 1]

### Evaluating the Algorithm from sklearn

### From the confusion matrix, our alogrithm misclassified only 24 out of 107. This is 77% accuracy

In [64]:
from sklearn.metrics import classification_report, confusion_matrix
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[33 13]
 [11 50]]
 precision recall f1-score support

 0 0.75 0.72 0.73 46
 1 0.79 0.82 0.81 61

 micro avg 0.78 0.78 0.78 107
 macro avg 0.77 0.77 0.77 107
weighted avg 0.77 0.78 0.78 107

### Conclusion: It shows that people with heart disease tend to be older, and have higher blood pressure, higher cholesterol levels, deeper and more widespread ST etc., than people without the disease.